ui/library/src/context/PageRenderContext.ts

```diff
@@ -139,13 +154,26 @@ export function usePageRenderContextProps({
   );
 
   React.useEffect(() => {
-    for (const pageNumber of visiblePageRatios.keys()) {
-      if (pageRenderStates.has(pageNumber)) {
-        continue;
-      }
+    const visiblePages = [...visiblePageRatios.keys()];
+    if (
+      !pdfDocProxy ||
+      [...pageRenderStates.keys()].length === pdfDocProxy.numPages ||
+      visiblePages.length === 0
+    ) {
+      return;
+    }
+
+    const visiblePagesNeighbors = [
```

Comment:
@ericmarsh995 i would love to abstract this line from 166 - 171 into maybe a class itself where we can unit test it correctly.